File: 04-Binary-Search-Tree/Code-Java/BST.java
package binarySearchTree;

import java.util.LinkedList;

/**
 * 二叉搜索树
 */
public class BST<Key extends Comparable<Key>, Value> {
    // 二叉搜索树的根节点
    private Node root;

   // 返回二分搜索树的节点个数
    public int size() {
        return size(root);
    }

    // 返回二分搜索树是否为空
    public boolean isEmpty() {
        return root != null && root.N == 0;
    }

    // 向二分搜索树中插入一个新的(key, Value)数据对
    public void insert(Key key, Value val) {
        root = insert(root, key, val);
    }

    // 在二分搜索树中搜索键 key 所对应的值。如果这个值不存在, 则返回 null
    public Value search(Key key) {
        return search(root, key);
    }

    // 深度优先遍历-前中后序遍历
    public void preOrder() {
        preOrder(root);
    }

    public void inOrder() {
        inOrder(root);
    }

    public void postOrder() {
        postOrder(root);
    }

    // 广度优先遍历
    public void levelOrder() {
        // 使用 LinkedList 作为辅助队列
        LinkedList<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            Node node = q.remove();
            System.out.println(node.key);

            if (node.left != null)
                q.add(node.left);
            if (node.right != null)
                q.add(node.right);
        }
    }

    // 返回二分搜索树的最小的键值
    public Key getMin() {
        Node minNode = getMin(root);
        return minNode.key;
    }

    // 返回二分搜索树的最大的键值
    public Key getMax() {
        Node maxNode = getMax(root);
        return maxNode.key;
    }

    // 删除二分搜索树中最小值所在节点
    public void removeMin() {
        if (root != null)
            root = removeMin(root);
    }

    // 删除二分搜索树中最大值所在节点
    public void removeMax() {
        if (root != null)
            root = removeMax(root);
    }

    // 从二分搜索树中删除键值为 key 的节点
    public void remove(Key key) {
        root = remove(root, key);
    }

    // 找到排名为 k 的键，即有 k 个小于它的键，排名从0开始
    public Key select(int k) {
        return select(root, k).key;
    }

    // 找到 key 的排名，排名从0开始
    public int rank(Key key) {
        return rank(root, key);
    }


    /*
    辅助函数
     */

    // 返回二分搜索树某节点为根节点的子树中节点的个数（含该节点）
    private int size(Node node) {
        if (node == null)
            return 0;
        return node.N;
    }

    // 如果 key 存在于 node 为根节点的子树中就更新它的值
    // 不存在就将 key-value 键值对作为新节点插入到子树中
    private Node insert(Node node, Key key, Value val) {
        if (node == null)
            return new Node(key, val, 1);
        int cmp = node.key.compareTo(key);
        if (cmp == 0)
            node.val = val;
        else if (cmp < 0)
            node.right = insert(node.right, key, val);
        else
            node.left = insert(node.left, key, val);
        node.N = size(node.left) + size(node.right) + 1;
        return node;
    }

    // 在以 node 为根节点的子树中查找并返回 key 所对应的值
    // 如果找不到返回 null
    private Value search(Node node, Key key) {
        if (node == null)
            return null;
        int cmp = node.key.compareTo(key);
        if (cmp == 0)
            return node.val;
        else if (cmp < 0)
            return search(node.right, key);
        else
            return search(node.left, key);
    }

    // 深度优先遍历
    // 对以 node 为根的二叉搜索树进行基于递归的前序遍历
    private void preOrder(Node node) {
        if (node != null) {
            System.out.println(node.key);
            preOrder(node.left);
            preOrder(node.right);
        }
    }

    // 深度优先遍历
    // 对以 node 为根的二叉搜索树进行基于递归的中序遍历
    private void inOrder(Node node) {
        if (node != null) {
            inOrder(node.left);
            System.out.println(node.key);
            inOrder(node.right);
        }
    }

    // 深度优先遍历
    // 对以 node 为根的二叉搜索树进行基于递归的后序遍历
    private void postOrder(Node node) {
        if (node != null) {
            postOrder(node.left);
            postOrder(node.right);
            System.out.println(node.key);
        }
    }

    // 返回以 node 为根的二分搜索树的最小键值所在的节点
    private Node getMin(Node node) {
        if (node.left == null)
            return node;
        return getMin(node.left);
    }

    // 返回以 node 为根的二分搜索树的最大键值所在的节点
    private Node getMax(Node node) {
        if (node.right == null)
            return node;
        return getMax(node.right);
    }

    // 删除以 node 为根的二分搜索树中的最小节点
    // 返回删除节点后新的二分搜索树的根
    private Node removeMin(Node node) {
        if (node.left == null) {
            Node nodeRight = node.right;
            node.right = null;
            return nodeRight;
        }
        node.left = removeMin(node.left);
        shiftN(node);
        return node;
    }

    // 删除以 node 为根的二分搜索树中的最大节点
    // 返回删除节点后新的二分搜索树的根
    private Node removeMax(Node node) {
        if (node.right == null) {
            Node nodeLeft = node.left;
            node.left = null;
            return nodeLeft;
        }
        node.right = removeMax(node.right);
        shiftN(node);
        return node;
    }

    // 删除以 node 为根的二分搜索树中键值为 key 的节点, 递归算法
    // 返回删除节点后新的二分搜索树的根
    private Node remove(Node node, Key key) {
        if (node == null)
            return null;

        int cmp = node.key.compareTo(key);
        if (cmp > 0) {
            node.left = remove(node.left, key);
            shiftN(node);
            return node;
        } else if (cmp < 0) {
            node.right = remove(node.right, key);
            shiftN(node);
            return node;
        } else {
            if (node.left == null) { //待删除节点左子树为空
                Node nodeRight = node.right;
                node.right = null;
                return nodeRight;
            } else if (node.right == null) { //待删除节点右子树为空的情况
                Node nodeLeft = node.left;
                node.left = null;
                return nodeLeft;
            } else {
                // 待删除节点左右子树均不为空
                // 找到比待删除节点大的最小节点, 即待删除节点右子树的最小节点
                // 用这个节点顶替待删除节点的位置
                Node t = node;
                node = getMin(t.right);
                node.right = removeMin(t.right);
                node.left = t.left;
                shiftN(node);
                return node;
            }
        }
    }

    // 使各节点以本节点为根节点的子树中节点的个数正确
    private void shiftN(Node node) {
        if (node.left == null && node.right == null)
            node.N = 1;
        else if (node.left == null)
            node.N = node.right.N + 1;
        else if (node.right == null)
            node.N = node.left.N + 1;
        else node.N = node.left.N + node.right.N + 1;
    }

    // 返回排名为 k 的节点
    private Node select(Node node, int k) {
        if (node == null)
            return null;
        int t = size(node.left);
        if (t > k)
            return select(node.left, k);
        else if (t < k)
            return select(node.right, k - t - 1);
        else
            return node;
    }

    // 返回以 node 以根节点的子树中小于 node.key 的数量
    private int rank(Node node, Key key) {
        if (node == null)
            return 0;
        int cmp = key.compareTo(node.key);
        if (cmp < 0)
            return rank(node.left, key);
        else if (cmp > 0)
            return rank(node.right, key) + size(node.left) + 1;
        else
            return size(node.left);
    }

    private class Node {
        private Key key; // 键
        private Value val; // 值
        private Node left, right; // 指向子节点的链接
        private int N; // 该节点为根节点的子树中节点的个数（含该节点）

        private Node(Key key, Value val, int N) {
            this.key = key;
            this.val = val;
            this.N = N;
        }
    }
}

File: 04-Binary-Search-Tree/Code-Python/BST.py
"""
二叉搜索树
"""
import queue


class BST:
    class Node:
        """树节点"""

        def __init__(self, data, left=None, right=None):
            self.data = data
            self.left = left
            self.right = right

    def __init__(self, source_collection=None):
        self.size = 0  # 二叉搜索树中节点的个数
        self.root = None  # 根节点
        if source_collection:
            for item in source_collection:
                self.insert(item)

    def __iter__(self):
        if self.root is not None:
            lst = list()
            lst.append(self.root)
            while lst:
                node = lst.pop()
                yield node.data
                if node.right is not None:
                    lst.append(node.right)
                if node.left is not None:
                    lst.append(node.left)

    def __str__(self):
        def recurse(node, level):
            s = ''
            if node is not None:
                s += recurse(node.right, level + 1)
                s += '| ' * level
                s += str(node.data) + '\n'
                s += recurse(node.left, level + 1)
            return s

        return recurse(self.root, 0)

    def __len__(self):
        return self.size

    def __eq__(self, other):
        if self is other:
            return True
        if type(self) != type(other) or len(self) != len(other):
            return False
        other_iter = iter(other)
        for item in self:
            if item != next(other_iter):
                return False
        return True

    def __add__(self, other):
        for item in other:
            self.insert(item)
        return self

    def is_empty(self):
        """列表为空返回 False，否则返回 True"""
        return self.size == 0

    def clear(self):
        """将列表清空"""
        self.root = None
        self.size = 0

    def search(self, item):
        """在二叉搜索树中搜索 item。如果存在, 返回第一个匹配的项，否则返回 None"""

        def recurse(node):
            if node is None:
                return None
            elif item == node.data:
                return node.data
            elif item < node.data:
                return recurse(node.left)
            else:
                return recurse(node.right)

        return recurse(self.root)

    def pre_order(self):
        """深度优先遍历-前序遍历
        把遍历的每一项添加到列表中，并返回该列表的迭代器
        """
        lst = list()

        def recurse(node):
            if node is not None:
                lst.append(node.data)
                recurse(node.left)
                recurse(node.right)

        recurse(self.root)
        return iter(lst)

    def in_order(self):
        """深度优先遍历-中序遍历
        把遍历的每一项添加到列表中，并返回该列表的迭代器
        """
        lst = list()

        def recurse(node):
            if node is not None:
                recurse(node.left)
                lst.append(node.data)
                recurse(node.right)

        recurse(self.root)
        return iter(lst)

    def post_order(self):
        """深度优先遍历-后序遍历
        把遍历的每一项添加到列表中，并返回该列表的迭代器
        """
        lst = list()

        def recurse(node):
            if node is not None:
                recurse(node.left)
                recurse(node.right)
                lst.append(node.data)

        recurse(self.root)
        return iter(lst)

    def level_order(self):
        """广度优先遍历
        把遍历的每一项添加到列表中，并返回该列表的迭代器
        """
        lst = list()
        q = queue.Queue()
        q.put(self.root)
        while not q.empty():
            node = q.get()
            lst.append(node.data)
            if node.left:
                q.put(node.left)
            if node.right:
                q.put(node.right)
        return lst

    def insert(self, item):
        """将一个项插入到二叉搜索树的适当位置
        如果 item 与某一项的值相等，插入到这一项的右子树中
        """

        def recurse(node):
            if item < node.data:
                if node.left is None:
                    node.left = BST.Node(item)
                else:
                    recurse(node.left)
            elif node.right is None:
                node.right = BST.Node(item)
            else:
                recurse(node.right)

        if self.is_empty():
            self.root = BST.Node(item)
        else:
            recurse(self.root)
        self.size += 1

    def get_min(self):
        """获取二叉搜索树中最小项"""

        def recurse(node):
            if node.left is None:
                return node.data
            return recurse(node.left)

        return recurse(self.root)

    def get_max(self):
        """获取二叉搜索树中最大项"""

        def recurse(node):
            if node.right is None:
                return node.data
            return recurse(node.right)

        return recurse(self.root)

    def remove_min(self):
        """删除二叉搜索树中最小项"""
        if self.size == 0:
            raise KeyError("二叉搜索树为空！")

        def recurse(node):
            if node.left.left is None:
                node.left = node.left.right
            else:
                recurse(node.left)

        recurse(self.root)
        self.size -= 1

    def remove_max(self):
        """删除二叉搜索树中最大项"""
        if self.size == 0:
            raise KeyError("二叉搜索树为空！")

        def recurse(node):
            if node.right.right is None:
                node.right = node.right.left
            else:
                recurse(node.right)

        recurse(self.root)
        self.size -= 1

    def remove(self, item):
        """在二叉搜索树中删除项 item。"""
        if item not in self:
            raise KeyError("需要移除的项不存在！")

        def recurse(node):
            if item == node.data:
                if node.left is None:  # 左子树为空
                    return node.right
                elif node.right is None:  # 右子树为空
                    return node.left
                else:  # 左右子树均不为空
                    new_node = self._pop_right_min(node)
                    new_node.left, new_node.right = node.left, node.right
                    return new_node
            elif item < node.data:
                node.left = recurse(node.left)
                return node
            else:
                node.right = recurse(node.right)
                return node

        self.root = recurse(self.root)

    @staticmethod
    def _pop_right_min(node):
        if node.right.left is None:
            pop_node = node.right
            node.right = node.right.right
        else:
            node = node.right
            while node.left.left is not None:
                node = node.left
            pop_node = node.left
            node.left = node.left.right
        return pop_node


bst = BST([10, 6, 3, 4, 5, 8, 9, 7, 17, 14, 19, 12, 18, 21, 24, 16])
# bst = BST([10, 6, 3, 4, 5])
print(bst)
# bst.remove_min()
# print(bst)
# bst.remove_max()
# print(bst)
# bst.remove(3)
# print(bst)

# for i in bst:
#     print(i)

# bst1 = BST([7, 17, 14, 19, 12])
# bst += bst1
# print(bst)

# print(list(bst.level_order()))
bst.remove_min()
print(bst)
